### ISSUE TITLE
Switch to use `events.EventRecorder` instead of `record.EventRecorder`

### ISSUE THREAD

**Issue by grzesuav** [kubernetes-sigs/controller-runtime#2141]
Controller runtime uses older interface, would be good to utilize new one

**Comment by alculquicondor** [comment]
/reopen

@alvaroaleman are there any considerations with regards to backwards-compatibility? Should we support both?

**Comment by alvaroaleman** [comment]
No, lets not support both. Is the old interface deprecated? Just for my understand as I have zero background here, what are reasons to use the `events.EventRecorder` over the `record.EventRecorder`?

**Comment by alculquicondor** [comment]
The one in `events` allows to provide extra information.

For example, you can have a secondary object and an action https://pkg.go.dev/k8s.io/client-go@v0.28.4/tools/events#EventRecorder

As a result, by migrating, we are forcing any users of `GetEventRecorderFor` to change their code.

**Comment by alvaroaleman** [comment]
Ah. So generally we allow breaking changes in minor version, so if we trade it for some benefit, that seems fine to me.

**Comment by mimowo** [comment]
/reopen 
To revisit the discussion and see if we could consider supporting both interfaces? This way applications using the controller-runtime would not need to change code on upgrade, while some applications could start using the new API. For example. the new interface could be exposed by a new function, such as `NewEventRecorderFor`.

WDYT @alvaroaleman ?

**Comment by alvaroaleman** [comment]
I am in favor of breaking it rather than supporting both interfaces

**Comment by mimowo** [comment]
I see, does it get significantly more work to support both?

If both are supported, then the upgrade of `controller-runtime` could be friction-less for apps, and would not require a bump of the major version of controller-runtime (I imagine). Then, we could deprecate the old recorder only when the old type is deprecated upstream k8s.

**Comment by mimowo** [comment]
This sounds innocent to me, is the underneath much implementation that would need to be repeated?

**Comment by sbueringer** [comment]
>  and would not require a bump of the major version of controller-runtime

controller-runtime is still using 0., so this will go into a minor version in any case.

> This sounds innocent to me, is the underneath much implementation that would need to be repeated?

Good question, someone will have to do some research :)

**Comment by mimowo** [comment]
/reopen
We have an assignee per https://github.com/kubernetes-sigs/controller-runtime/issues/2141#issuecomment-2272974128

**Comment by troy0820** [comment]
If we are going to move to the new interface (this doesn't look like it's planned for the next release) we can deprecate the old interface `recorder.EventRecorder` and by release-0.20 remove it in favor of the new `events.EventRecorder`.  

This would give time to people needing to migrate to the new interface. The breaking change is inevitable, but it's not really a drop in replacement with the function signature being different.

**Comment by clebs** [comment]
Hi folks,

I have been investigating how to make this migration possible. From the `EventRecorder` perspective it is faily simple.
The tricky part is migrating to the new `EventBroadcaster`, which is quite different from the old one.

Since this change affects the entire codebase, I would also argue against keeping both APIs at the same time because it would mean we need duplication across the board and on tests.

I have a draft PR with most of the changes done to see if this is the right direction or we should go for a completely different approach: #3262

I would appreciate feedback especially in how broadcasters are injected for testing and the general plumbing throughout the codebase.
I tried to make the injection more straightforward and with less indirection now that the constructor has changed for broadcasters.

There are also some open challenges where I left comments (e.g. `resourcelock.ResourceLockConfig` still uses the old API).

**Comment by clebs** [comment]
The fact that client-go's [ResourceLockConfig](https://github.com/kubernetes/client-go/blob/06fcc8a0753eec9e896c5f96985fc6595ef74391/tools/leaderelection/resourcelock/interface.go#L68) (used in leader election) is still using the old events API forces us to support both for now.
Therefore I have added a way to somewhat still support the old API with minimal duplication.

**Comment by pmalek** [comment]
👋 

We just hit this after trying to upgrade to 0.23.0 (through a transitive dependency in https://github.com/Kong/kong-operator/pull/3105)

"New" API - https://github.com/kubernetes/client-go/blob/v0.34.3/tools/internal/events/interfaces.go#L28-L44 - does not allow annotating the published event. 

The "old" API - https://github.com/kubernetes/client-go/blob/v0.34.3/tools/record/event.go#L110-L111 - did allow it.

I don't see anyone mentioning it so it must be the first time this is mentioned. Is there an option to annotate an event? If not, is there an issue upstream to add something similar to `AnnotatedEventf` to the new `events` API?

**Comment by sbueringer** [comment]
> Is there an option to annotate an event?

Not sure, I also don't see one. But the question might not be specific to client-go. Maybe the question is if the new events API in general still supports annotations (I didn't take a closer look)

> If not is there an issue upstream to add something similar to AnnotatedEventf to the new events API?

Not sure. Might be worth creating one or just asking in Slack #sig-apimachinery first. Maybe this is by design?

**Comment by clebs** [comment]
There is no longer way to annotate events with the new API.
I asked on #sig-api-machinery when i was working on this but got no answer.

Only thing I see as posible substitute is using the new action field for that purpose maybe?

**Comment by sbueringer** [comment]
Might be explained in the kep for the new events API

**Comment by clebs** [comment]
Good point, will see if I can review it when I have some time.

**Comment by sbueringer** [comment]
> 👋
> 
> We just hit this after trying to upgrade to 0.23.0 (through a transitive dependency in [Kong/kong-operator#3105](https://github.com/Kong/kong-operator/pull/3105))
> 
> "New" API - [kubernetes/client-go@`v0.34.3`/tools/internal/events/interfaces.go#L28-L44](https://github.com/kubernetes/client-go/blob/v0.34.3/tools/internal/events/interfaces.go?rgh-link-date=2026-01-22T11%3A02%3A59.000Z#L28-L44) - does not allow annotating the published event.
> 
> The "old" API - [kubernetes/client-go@`v0.34.3`/tools/record/event.go#L110-L111](https://github.com/kubernetes/client-go/blob/v0.34.3/tools/record/event.go?rgh-link-date=2026-01-22T11%3A02%3A59.000Z#L110-L111) - did allow it.
> 
> I don't see anyone mentioning it so it must be the first time this is mentioned. Is there an option to annotate an event? If not, is there an issue upstream to add something similar to `AnnotatedEventf` to the new `events` API?

@pmalek I think this will be addressed in https://github.com/kubernetes-sigs/controller-runtime/pull/3509

**Comment by adri1197** [comment]
Regarding annotation support: there's an upstream PR to add `AnnotatedEventf` to the new events API in client-go kubernetes/kubernetes#138103 (https://github.com/kubernetes/kubernetes/pull/138103). The target release is K8s v1.37. It introduces a separate `AnnotatedEventRecorder` interface (to avoid breaking existing `EventRecorder` implementors) and wires annotations through to `ObjectMeta.Annotations` on the resulting eventsv1.Event.

The controller-runtime counterpart is #3509, which adds `AnnotatedEventRecorderProvider` as an opt-in interface (type assertion, no breaking changes to Cluster or Manager interfaces)
